### PR TITLE
Basic Downloader

### PR DESCRIPTION
**Pull request by faultydev**
This fork implements a basic downloading tool in python.  
It also implements a simple protocol to be used by the graphic side of the installer.

### CODE REVIEW

**Comment by WillKenzie** [comment]
Notice: Offical Cybfi Response enclosed.

The daemon is split into ideally one major part per operating system. Again, ideally, these parts will be compatible with each other regardless of the system the user is on. This was executed perfectly by yourself in the creation of the installer, however, it doesn't appear that in the downloading of the Daemon there is a separation between systems. Is your proposal to then download the macOS/Linux *and* Windows versions of the daemon in the same ZIP? Or do you plan to contribute to some form of a universal method of running it across platforms? Note that we don't want a Python dependency (at least not that the user can notice, if we must do a Python dependency then it will install just the parts it needs from Python in a special directory as to not conflict with other versions of Python)

In short, before we can merge this pull request, we must know how to make package.zip compatible with the installer, or how to make the installer compatible with a new format of package.zip.

My official recommendation is to segment the downloads into "Windows.zip" and "Unix.zip" and have the installer recognize the OS and download the appropriate install. If you'd rather submit a pull request to https://github.com/cybfi/daemon and support a way to run the entire daemon independent of OS, that is fine as well.

Will Not Merge (as of November 5th @ 2:44PM CST)

**Comment by faultydev** [comment]
Will, the PR states: "**WIP**: Basic Downloader."
Not looking for it to get merged *right now*, just looking for feedback ;)
I'll make the installer install process compatible with multiple operating systems in the future; likely another PR. This PR only houses an universal downloader that if you think is alright to base off on I can build on top of :p 

Cheers, Didier

**Comment by WillKenzie** [comment]
I understand it is a work in progress, but I have a protocol to follow. I gave you feedback and suggestions, take them as you will. If you wouldn't mind holding onto the fork for a little while longer until what was mentioned is worked out, then I will be glad to merge when it's ready.

This is a great starting point and I'm thrilled that you wish to work on the repository. 

Sincerely,
Will

**Comment by faultydev** [comment]
Asking for review (you should add labels for this ;))

File: src/lib/helpers.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import os
+
+
+def envBYN(key: str, val=None):
+    if val is None:
+        val = os.getenv(key)
+    elif type(val) is not str:
+        val = str(val)
+    # envBYN (Binary Yes No) is a helper function to convert a string to a boolean
```
Comment (WillKenzie) [inline src/lib/helpers.py:9]: For simplicity's sake, move this to the top of the function to follow PEP standards.

Comment (faultydev) [reply src/lib/helpers.py:9]: fair, hold for commit :upside_down_face:

File: src/main.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import os
+import sys
+from zipfile import ZipFile
+
+from lib.Downloader import Downloader
+from lib.KnowWhere import KnowWhere
+
+base_daemon_url = "https://github.com/cybfi/daemon/releases/latest/download/"
```
Comment (WillKenzie) [inline src/main.py:8]: This returns 404 in testing.

**Review by WillKenzie — CHANGES_REQUESTED** [review]
There are a few commented changes that need to be made, especially with how the installer downloads files from Github.

File: src/main.py (hunk 2 above)
Comment (faultydev) [reply src/main.py:8]: It will when you make the files available :))))))

Comment (WillKenzie) [reply src/main.py:8]: go to https://github.com/Cybfi/daemon/releases/download/v0.0.1/main.py

now go to https://github.com/Cybfi/daemon/releases/download/latest/main.py

then try https://github.com/cybfi/daemon/releases/latest/download/main.py

Comment (faultydev) [reply src/main.py:8]: no fair, you have marked it a pre-release, so it doesn't work. :smirk:

Comment (WillKenzie) [reply src/main.py:8]: ... I will confer with some sources

Comment (WillKenzie) [reply src/main.py:8]: Confirmed. This is a mistake on my part. Changes will not be needed.

**Review by WillKenzie — APPROVED** [review]

**Comment by WillKenzie** [comment]
@faultydev Is this complete?

**Comment by faultydev** [comment]
> @faultydev Is this complete?

yup!

**Comment by WillKenzie** [comment]
Merging this as it is complete and approved.
